sphinx/ext/githubpages.py

```diff
@@ -22,11 +22,18 @@
 def create_nojekyll(app, env):
     # type: (Sphinx, BuildEnvironment) -> None
     if app.builder.format == 'html':
-        path = os.path.join(app.builder.outdir, '.nojekyll')
-        open(path, 'wt').close()
+        open(os.path.join(app.builder.outdir, '.nojekyll'), 'wt').close()
+
+        cname = app.config.githubpages_cname
+        if cname:
+            with open(os.path.join(app.builder.outdir, 'CNAME'), 'wt') as f:
+                # NOTE: don't write a trailing newline. The `CNAME` file that's
+                # auto-generated by the Github UI doesn't have one.
+                f.write(cname)
 
 
 def setup(app):
     # type: (Sphinx) -> Dict[str, Any]
     app.connect('env-updated', create_nojekyll)
+    app.add_config_value("githubpages_cname", "", "html")
```

Comment:
It seems like `html_baseurl` expects a string like `https://sphinx-doc.org`, but the CNAME wants a string like `sphinx-doc.org`. How can we convert the first form into the second form across Python versions? Ordinarily I would use `urllib`, but it looks like the API has changed significantly between Python 2 and 3; and doing some string manipulation like splitting on `/` seems suspect to me.

Reply:
Actually, I think we can't do this. We may want to set `html_baseurl` to something like `https://sphinx-doc.github.io/sphinx-doc/sphinx/`. In that case, we don't actually own the domain `github.io`, and we don't want to generate the `CNAME` file. It's also possible there's some use of `html_baseurl` that I haven't thought of yet, but it would be hard to change the behavior of this extension in the future. Therefore I think it's best to opt in explicitly using a setting specifically for the extension.

Reply:
Sorry for late response.

>How can we convert the first form into the second form across Python versions?

Since 2.0, we'll only support py3.5+. So you can use `urllib.parse` for such purpose.

>In that case, we don't actually own the domain `github.io`, and we don't want to generate the `CNAME` file.

We can test it with `hostname.endswith('.github.io')` simply.